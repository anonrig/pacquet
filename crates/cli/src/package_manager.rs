use std::{io, path::PathBuf};

use pacquet_diagnostics::{
    miette::{self, Diagnostic},
    thiserror::{self, Error},
};
use pacquet_npmrc::{current_npmrc, Npmrc};
use pacquet_package_json::PackageJson;

use crate::package_cache::PackageCache;

#[derive(Error, Debug, Diagnostic)]
pub enum AutoImportError {
    #[error("cannot create directory at {dirname:?}: {error}")]
    CreateDir {
        dirname: PathBuf,
        #[source]
        error: io::Error,
    },
    #[error("fail to create a link from {from:?} to {to:?}: {error}")]
    CreateLink {
        from: PathBuf,
        to: PathBuf,
        #[source]
        error: io::Error,
    },
}

#[derive(Error, Debug, Diagnostic)]
#[non_exhaustive]
pub enum PackageManagerError {
    #[error(transparent)]
    #[diagnostic(transparent)]
    Tarball(#[from] pacquet_tarball::TarballError),

    #[error(transparent)]
    #[diagnostic(transparent)]
    PackageJson(#[from] pacquet_package_json::PackageJsonError),

    #[error(transparent)]
    #[diagnostic(transparent)]
    Registry(#[from] pacquet_registry::RegistryError),

    #[error(transparent)]
    #[diagnostic(code(pacquet_package_manager::io_error))]
    Io(#[from] io::Error),

    #[error(transparent)]
    #[diagnostic(transparent)]
    AutoImport(#[from] AutoImportError),
}

pub struct PackageManager {
    pub config: Box<Npmrc>,
    pub package_json: Box<PackageJson>,
    pub http_client: Box<reqwest::Client>,
    pub(crate) package_cache: PackageCache,
}

impl PackageManager {
    pub fn new<P: Into<PathBuf>>(package_json_path: P) -> Result<Self, PackageManagerError> {
        Ok(PackageManager {
            config: Box::new(current_npmrc()),
            package_json: Box::new(PackageJson::create_if_needed(package_json_path.into())?),
            http_client: Box::new(reqwest::Client::new()),
            package_cache: PackageCache::new(),
        })
    }
}
