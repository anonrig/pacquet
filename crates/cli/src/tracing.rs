use std::{str::FromStr, sync::atomic::AtomicBool};

use tracing::Level;
use tracing_subscriber::{fmt::format::FmtSpan, layer::Filter, EnvFilter, Layer};

static IS_TRACING_ENABLED: AtomicBool = AtomicBool::new(false);

// skip event because it's not useful for performance analysis
struct FilterEvent;

impl<S> Filter<S> for FilterEvent {
    fn enabled(
        &self,
        meta: &tracing::Metadata<'_>,
        _cx: &tracing_subscriber::layer::Context<'_, S>,
    ) -> bool {
        !meta.is_event()
    }
}
pub fn enable_tracing_by_env() {
    let trace_var = std::env::var("TRACE").ok();
    let is_enable_tracing = trace_var.is_some();

    if is_enable_tracing && !IS_TRACING_ENABLED.swap(true, std::sync::atomic::Ordering::SeqCst) {
        use tracing_subscriber::{fmt, prelude::*};
        let layers = generate_common_layers(trace_var);

        tracing_subscriber::registry()
            .with(layers)
            .with(fmt::layer().pretty().with_file(true).with_span_events(FmtSpan::CLOSE))
            .init();
        tracing::trace!("enable_tracing_by_env");
    }
}

fn generate_common_layers(
    trace_var: Option<String>,
) -> Vec<Box<dyn Layer<tracing_subscriber::Registry> + Send + Sync>> {
    let default_level = trace_var.as_ref().and_then(|var| Level::from_str(var).ok());

    let mut layers = vec![];
    if let Some(default_level) = default_level {
        layers.push(
            tracing_subscriber::filter::Targets::new()
                .with_targets(vec![("pacquet_tarball", default_level)])
                .boxed(),
        );
    } else {
        // SAFETY: we know that trace_var is `Ok(String)` now,
        // for the second unwrap, if we can't parse the directive, then the tracing result would be
        // unexpected, then panic is reasonable
        let env_layer = EnvFilter::builder()
      .with_regex(true)
      .parse(trace_var.expect("Should not be empty"))
      .expect("Parse tracing directive syntax failed,for details about the directive syntax you could refer https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html#directives");

        layers.push(env_layer.boxed());
    }
    layers
}
