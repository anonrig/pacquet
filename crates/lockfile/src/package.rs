use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::LockfilePeerDependencyMeta;

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct LockfilePackageResolution {
    integrity: String,
}

// Reference: https://github.com/pnpm/pnpm/blob/main/lockfile/lockfile-file/src/sortLockfileKeys.ts#L5
#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct LockfilePackage {
    resolution: LockfilePackageResolution,
    id: Option<String>,

    name: Option<String>,
    version: Option<String>,

    engines: Option<HashMap<String, String>>,
    cpu: Option<Vec<String>>,
    os: Option<Vec<String>>,
    // TODO: Add `libc`
    deprecated: Option<bool>,
    #[serde(alias = "hasBin")]
    has_bin: Option<bool>,
    // TODO: Add `prepare`
    #[serde(alias = "requiresBuild")]
    requires_build: Option<bool>,

    // TODO: Add `bundleDependencies`
    #[serde(alias = "peerDependencies")]
    peer_dependencies: Option<HashMap<String, String>>,
    #[serde(alias = "peerDependenciesMeta")]
    peer_dependencies_meta: Option<HashMap<String, LockfilePeerDependencyMeta>>,

    dependencies: Option<HashMap<String, String>>,
    optional_dependencies: Option<HashMap<String, String>>,

    #[serde(alias = "transitivePeerDependencies")]
    transitive_peer_dependencies: Option<Vec<String>>,
    dev: bool,
    optional: Option<bool>,
}
