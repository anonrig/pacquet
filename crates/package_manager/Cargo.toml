[package]
name                  = "pacquet_package_manager"
version               = "0.0.1"
publish               = false
authors.workspace     = true
description.workspace = true
edition.workspace     = true
homepage.workspace    = true
keywords.workspace    = true
license.workspace     = true
repository.workspace  = true

[dependencies]
pacquet_fs          = { workspace = true }
pacquet_lockfile    = { workspace = true }
pacquet_npmrc       = { workspace = true }
pacquet_registry    = { workspace = true }
pacquet_tarball     = { workspace = true }

derive_more   = { workspace = true }
futures-util  = { workspace = true }
pipe-trait    = { workspace = true }
rayon         = { workspace = true }
reflink-copy  = { workspace = true }
reqwest       = { workspace = true }
tracing       = { workspace = true }
miette        = { workspace = true }

[dev-dependencies]
node-semver       = { workspace = true }
pretty_assertions = { workspace = true }
tempfile          = { workspace = true }
tokio             = { workspace = true }
