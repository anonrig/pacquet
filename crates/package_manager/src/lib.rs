mod create_cas_files;
mod create_symlink_layout;
mod create_virtual_dir_by_snapshot;
mod create_virtual_store;
mod install_package_by_snapshot;
mod install_package_from_registry;
mod link_file;
mod symlink_package;

pub use create_cas_files::{create_cas_files, CreateCasFilesError};
pub use create_symlink_layout::create_symlink_layout;
pub use create_virtual_dir_by_snapshot::{CreateVirtualDirBySnapshot, CreateVirtualDirError};
pub use create_virtual_store::CreateVirtualStore;
pub use install_package_by_snapshot::{InstallPackageBySnapshot, InstallPackageBySnapshotError};
pub use install_package_from_registry::{
    InstallPackageFromRegistry, InstallPackageFromRegistryError,
};
pub use link_file::{link_file, LinkFileError};
pub use symlink_package::{symlink_package, SymlinkPackageError};
