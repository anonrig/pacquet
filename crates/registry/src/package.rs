use std::collections::HashMap;

use reqwest::Client;
use serde::{Deserialize, Serialize};

use crate::error::RegistryError;

pub enum PackageType {
    Dependency,
    DevDependency,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct PackageDistribution {
    pub integrity: String,
    #[serde(alias = "npm-signature")]
    pub npm_signature: Option<String>,
    pub shasum: String,
    pub tarball: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct PackageVersion {
    #[serde(alias = "_npmVersion")]
    pub npm_version: String,
    #[serde(alias = "_nodeVersion")]
    pub node_version: Option<String>,
    pub dist: PackageDistribution,
    pub dependencies: Option<HashMap<String, String>>,
    #[serde(alias = "devDependencies")]
    pub dev_dependencies: Option<HashMap<String, String>>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Package {
    pub name: String,
    #[serde(alias = "dist-tags")]
    dist_tags: HashMap<String, String>,
    pub versions: HashMap<String, PackageVersion>,
}

impl Package {
    pub async fn from_registry(
        client: &Client,
        package_url: &str,
    ) -> Result<Package, RegistryError> {
        Ok(client
            .get(package_url)
            .header("user-agent", "pacquet-cli")
            .header("content-type", "application/json")
            .send()
            .await?
            .json::<Package>()
            .await?)
    }

    pub fn get_latest_tag(&self) -> Result<&String, RegistryError> {
        self.dist_tags.get("latest").ok_or(RegistryError::MissingLatestTag(self.name.to_owned()))
    }

    pub fn get_latest_version(&self) -> Result<&PackageVersion, RegistryError> {
        let latest_tag = self.get_latest_tag()?;
        self.versions.get(latest_tag).ok_or(RegistryError::MissingVersionRelease(
            latest_tag.to_owned(),
            self.name.to_owned(),
        ))
    }

    pub fn get_tarball_url(&self) -> Result<&str, RegistryError> {
        Ok(self.get_latest_version()?.dist.tarball.as_str())
    }
}
