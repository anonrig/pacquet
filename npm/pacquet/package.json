{
  "name": "pacquet",
  "version": "0.0.1",
  "description": "experimental package manager for node.js",
  "keywords": [],
  "author": {
    "name": "Yagiz Nizipli",
    "email": "yagiz@nizipli.com"
  },
  "license": "MIT",
  "homepage": "https://github.com/anonrig/pacquet",
  "bugs": "https://github.com/anonrig/pacquet/issues",
  "repository": {
    "type": "git",
    "url": "https://github.com/anonrig/pacquet",
    "directory": "npm/pacquet"
  },
  "bin": "bin/pacquet",
  "engines": {
    "node": ">=18.*"
  },
  "files": [
    "bin/pacquet"
  ]
}