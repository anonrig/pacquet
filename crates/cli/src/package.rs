use crate::package_manager::PackageManagerError;
use pacquet_diagnostics::tracing;
use pacquet_npmrc::Npmrc;
use pacquet_package_manager::{create_cas_files, symlink_pkg};
use pacquet_registry::{Package, PackageVersion};
use pacquet_tarball::{download_tarball_to_store, Cache};
use reqwest::Client;
use std::{fmt::Display, path::Path, str::FromStr};

/// This function execute the following and returns the package
/// - retrieves the package from the registry
/// - extracts the tarball to global store directory (~/Library/../pacquet)
/// - links global store directory to virtual dir (node_modules/.pacquet/..)
///
/// symlink_path will be appended by the name of the package. Therefore,
/// it should be resolved into the node_modules folder of a subdependency such as
/// `node_modules/.pacquet/fastify@1.0.0/node_modules`.
pub async fn install_package_from_registry<Tag>(
    tarball_cache: &Cache,
    config: &'static Npmrc,
    http_client: &Client,
    name: &str,
    version_range: &str,
    symlink_path: &Path,
) -> Result<PackageVersion, PackageManagerError>
where
    Tag: FromStr + Display,
{
    // TODO: create a PackageTag enum with a parse function
    Ok(if let Ok(version) = version_range.parse::<Tag>() {
        let package_version =
            PackageVersion::fetch_from_registry(name, version, http_client, &config.registry)
                .await
                .map_err(PackageManagerError::Registry)?;
        internal_fetch(tarball_cache, http_client, &package_version, config, symlink_path).await?;
        package_version
    } else {
        let package = Package::fetch_from_registry(name, http_client, &config.registry)
            .await
            .map_err(PackageManagerError::Registry)?;
        let package_version = package.pinned_version(version_range).unwrap(); // TODO: propagate error for when no version satisfies range
        internal_fetch(tarball_cache, http_client, package_version, config, symlink_path).await?;
        package_version.clone()
    })
}

async fn internal_fetch(
    tarball_cache: &Cache,
    http_client: &Client,
    package_version: &PackageVersion,
    config: &'static Npmrc,
    symlink_path: &Path,
) -> Result<(), PackageManagerError> {
    let store_folder_name = package_version.to_virtual_store_name();

    // TODO: skip when it already exists in store?
    let cas_paths = download_tarball_to_store(
        tarball_cache,
        http_client,
        &config.store_dir,
        package_version.dist.integrity.as_ref().expect("has integrity field"),
        package_version.dist.unpacked_size,
        package_version.as_tarball_url(),
    )
    .await
    .map_err(PackageManagerError::Tarball)?;

    let save_path = config
        .virtual_store_dir
        .join(store_folder_name)
        .join("node_modules")
        .join(&package_version.name);

    let symlink_path = symlink_path.join(&package_version.name);

    tracing::info!(target: "pacquet::import", ?save_path, ?symlink_path, "Import package");

    create_cas_files(config.package_import_method, &save_path, &cas_paths)
        .map_err(PackageManagerError::CreateCasFiles)?;

    symlink_pkg(&save_path, &symlink_path).map_err(PackageManagerError::SymlinkPackage)?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use crate::package::install_package_from_registry;
    use node_semver::Version;
    use pacquet_npmrc::Npmrc;
    use pipe_trait::Pipe;
    use pretty_assertions::assert_eq;
    use std::fs;
    use std::path::Path;
    use tempfile::tempdir;

    fn create_config(store_dir: &Path, modules_dir: &Path, virtual_store_dir: &Path) -> Npmrc {
        Npmrc {
            hoist: false,
            hoist_pattern: vec![],
            public_hoist_pattern: vec![],
            shamefully_hoist: false,
            store_dir: store_dir.to_path_buf(),
            modules_dir: modules_dir.to_path_buf(),
            node_linker: Default::default(),
            symlink: false,
            virtual_store_dir: virtual_store_dir.to_path_buf(),
            package_import_method: Default::default(),
            modules_cache_max_age: 0,
            lockfile: false,
            prefer_frozen_lockfile: false,
            lockfile_include_tarball_url: false,
            registry: "https://registry.npmjs.com/".to_string(),
            auto_install_peers: false,
            dedupe_peer_dependents: false,
            strict_peer_dependencies: false,
            resolve_peers_from_workspace_root: false,
        }
    }

    #[tokio::test]
    pub async fn should_find_package_version_from_registry() {
        let store_dir = tempdir().unwrap();
        let modules_dir = tempdir().unwrap();
        let virtual_store_dir = tempdir().unwrap();
        let config: &'static Npmrc =
            create_config(store_dir.path(), modules_dir.path(), virtual_store_dir.path())
                .pipe(Box::new)
                .pipe(Box::leak);
        let http_client = reqwest::Client::new();
        let symlink_path = tempdir().unwrap();
        let package = install_package_from_registry::<Version>(
            &Default::default(),
            config,
            &http_client,
            "fast-querystring",
            "1.0.0",
            symlink_path.path(),
        )
        .await
        .unwrap();

        assert_eq!(package.name, "fast-querystring");
        assert_eq!(
            package.version,
            Version { major: 1, minor: 0, patch: 0, build: vec![], pre_release: vec![] }
        );

        let virtual_store_path = virtual_store_dir
            .path()
            .join(package.to_virtual_store_name())
            .join("node_modules")
            .join(&package.name);
        assert!(virtual_store_path.is_dir());

        // Make sure the symlink is resolving to the correct path
        assert_eq!(
            fs::read_link(symlink_path.path().join(&package.name)).unwrap(),
            virtual_store_path
        );
    }
}
