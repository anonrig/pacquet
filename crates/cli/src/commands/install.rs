use crate::package::{install_package_from_registry, install_single_package_to_virtual_store};
use crate::package_manager::{PackageManager, PackageManagerError};
use async_recursion::async_recursion;
use clap::Parser;
use futures_util::future;
use pacquet_diagnostics::tracing;
use pacquet_lockfile::{
    DependencyPath, Lockfile, PackageSnapshot, PkgNameVerPeer, RootProjectSnapshot,
};
use pacquet_package_json::DependencyGroup;
use pacquet_registry::PackageVersion;
use pipe_trait::Pipe;
use std::{collections::HashMap, io::ErrorKind};

#[derive(Parser, Debug)]
pub struct InstallCommandArgs {
    /// pacquet will not install any package listed in devDependencies and will remove those insofar
    /// they were already installed, if the NODE_ENV environment variable is set to production.
    /// Use this flag to instruct pacquet to ignore NODE_ENV and take its production status from this
    /// flag instead.
    #[arg(short = 'P', long = "prod")]
    pub prod: bool,
    /// Only devDependencies are installed and dependencies are removed insofar they were
    /// already installed, regardless of the NODE_ENV.
    #[arg(short = 'D', long = "dev")]
    pub dev: bool,
    /// optionalDependencies are not installed.
    #[arg(long = "no-optional")]
    pub no_optional: bool,
}

impl InstallCommandArgs {
    /// Convert the command arguments to an iterator of [`DependencyGroup`]
    /// which filters the types of dependencies to install.
    fn dependency_groups(&self) -> impl Iterator<Item = DependencyGroup> {
        let &InstallCommandArgs { prod, dev, no_optional } = self;
        let has_both = prod == dev;
        let has_prod = has_both || prod;
        let has_dev = has_both || dev;
        let has_optional = !no_optional;
        std::iter::empty()
            .chain(has_prod.then_some(DependencyGroup::Default))
            .chain(has_dev.then_some(DependencyGroup::Dev))
            .chain(has_optional.then_some(DependencyGroup::Optional))
    }
}

impl PackageManager {
    /// Install dependencies of a dependency.
    ///
    /// This function is used by [`PackageManager::install`] without a lockfile.
    #[async_recursion]
    async fn install_dependencies_from_registry(&self, package: &PackageVersion) {
        let node_modules_path = self
            .config
            .virtual_store_dir
            .join(package.to_virtual_store_name())
            .join("node_modules");

        tracing::info!(target: "pacquet::install", node_modules = ?node_modules_path, "Start subset");

        package
            .dependencies(self.config.auto_install_peers)
            .map(|(name, version_range)| async {
                let dependency = install_package_from_registry(
                    &self.tarball_cache,
                    self.config,
                    &self.http_client,
                    name,
                    version_range,
                    &node_modules_path,
                )
                .await
                .unwrap();
                self.install_dependencies_from_registry(&dependency).await;
            })
            .pipe(future::join_all)
            .await;

        tracing::info!(target: "pacquet::install", node_modules = ?node_modules_path, "Complete subset");
    }

    /// Generate filesystem layout for the virtual store at `node_modules/.pacquet`.
    async fn create_virtual_store(
        &self,
        packages: &Option<HashMap<DependencyPath, PackageSnapshot>>,
    ) {
        let Some(packages) = packages else {
            todo!("check project_snapshot, error if it's not empty, do nothing if empty");
        };
        packages
            .iter()
            .map(|(dependency_path, package_snapshot)| async move {
                install_single_package_to_virtual_store(
                    &self.tarball_cache,
                    &self.http_client,
                    self.config,
                    dependency_path,
                    package_snapshot,
                )
                .await
                .unwrap();
            })
            .pipe(future::join_all)
            .await;
    }

    /// Create symlinks for the direct dependencies.
    ///
    /// If package `foo@x.y.z` is declared as a dependency in `package.json`,
    /// symlink `foo -> .pacquet/foo@x.y.z/node_modules/foo` shall be created
    /// in the `node_modules` directory.
    async fn link_direct_dependencies(
        &self,
        project_snapshot: &RootProjectSnapshot,
        args: &InstallCommandArgs,
    ) {
        let RootProjectSnapshot::Single(project_snapshot) = project_snapshot else {
            panic!("Monorepo is not yet supported");
        };

        // TODO: parallelize this, either by tokio or rayon
        for (name, spec) in project_snapshot.dependencies_by_groups(args.dependency_groups()) {
            let custom_registry = None; // assuming all registries are default registries (custom registry is not yet supported)
            let package_specifier = PkgNameVerPeer::new(name.to_string(), spec.version.clone());
            let dependency_path = DependencyPath { custom_registry, package_specifier };
            let virtual_store_name = dependency_path.to_virtual_store_name();
            // NOTE: symlink target in pacquet is absolute yet in pnpm is relative
            // TODO: change symlink target to relative
            let symlink_target = self
                .config
                .virtual_store_dir
                .join(virtual_store_name)
                .join("node_modules")
                .join(name);
            let symlink_path = self.config.modules_dir.join(name);
            if let Err(error) = crate::fs::symlink_dir(&symlink_target, &symlink_path) {
                match error.kind() {
                    ErrorKind::AlreadyExists => {},
                    _ => panic!("Failed to create symlink at {symlink_path:?} to {symlink_target:?}: {error}"), // TODO: proper error propagation
                }
            }
        }
    }

    /// Jobs of the `install` command.
    pub async fn install(&self, args: &InstallCommandArgs) -> Result<(), PackageManagerError> {
        tracing::info!(target: "pacquet::install", "Start all");

        match (self.config.lockfile, &self.lockfile) {
            (false, _) => {
                self.package_json
                    .dependencies(args.dependency_groups())
                    .map(|(name, version_range)| async move {
                        let dependency = install_package_from_registry(
                            &self.tarball_cache,
                            self.config,
                            &self.http_client,
                            name,
                            version_range,
                            &self.config.modules_dir,
                        )
                        .await
                        .unwrap();
                        self.install_dependencies_from_registry(&dependency).await;
                    })
                    .pipe(future::join_all)
                    .await;
            }
            (true, None) => {
                unimplemented!();
            }
            (true, Some(lockfile)) => {
                let Lockfile { lockfile_version, project_snapshot, packages, .. } = lockfile;
                assert_eq!(lockfile_version.major, 6); // compatibility check already happens at serde, but this still helps preventing programmer mistakes.

                assert!(
                    self.config.prefer_frozen_lockfile,
                    "Non frozen lockfile is not yet supported",
                );

                future::join(
                    self.create_virtual_store(packages),
                    self.link_direct_dependencies(project_snapshot, args),
                )
                .await;
            }
        }

        tracing::info!(target: "pacquet::install", "Complete all");
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use std::env;

    use crate::commands::install::InstallCommandArgs;
    use crate::fs::get_all_folders;
    use crate::package_manager::PackageManager;
    use pacquet_npmrc::Npmrc;
    use pacquet_package_json::{DependencyGroup, PackageJson};
    use pretty_assertions::assert_eq;
    use tempfile::tempdir;

    #[test]
    fn install_args_to_dependency_groups() {
        use DependencyGroup::{Default, Dev, Optional};
        let create_list = |args: InstallCommandArgs| args.dependency_groups().collect::<Vec<_>>();

        // no flags -> prod + dev + optional
        assert_eq!(
            create_list(InstallCommandArgs { prod: false, dev: false, no_optional: false }),
            [Default, Dev, Optional],
        );

        // --prod -> prod + optional
        assert_eq!(
            create_list(InstallCommandArgs { prod: true, dev: false, no_optional: false }),
            [Default, Optional],
        );

        // --dev -> dev + optional
        assert_eq!(
            create_list(InstallCommandArgs { prod: false, dev: true, no_optional: false }),
            [Dev, Optional],
        );

        // --no-optional -> prod + dev
        assert_eq!(
            create_list(InstallCommandArgs { prod: false, dev: false, no_optional: true }),
            [Default, Dev],
        );

        // --prod --no-optional -> prod
        assert_eq!(
            create_list(InstallCommandArgs { prod: true, dev: false, no_optional: true }),
            [Default],
        );

        // --dev --no-optional -> dev
        assert_eq!(
            create_list(InstallCommandArgs { prod: false, dev: true, no_optional: true }),
            [Dev],
        );

        // --prod --dev -> prod + dev + optional
        assert_eq!(
            create_list(InstallCommandArgs { prod: true, dev: true, no_optional: false }),
            [Default, Dev, Optional],
        );

        // --prod --dev --no-optional -> prod + dev
        assert_eq!(
            create_list(InstallCommandArgs { prod: true, dev: true, no_optional: true }),
            [Default, Dev],
        );
    }

    #[tokio::test]
    pub async fn should_install_dependencies() {
        let dir = tempdir().unwrap();
        let current_directory = env::current_dir().unwrap();
        env::set_current_dir(dir.path()).unwrap();

        let package_json_path = dir.path().join("package.json");
        let mut package_json = PackageJson::create_if_needed(package_json_path.clone()).unwrap();

        package_json.add_dependency("is-odd", "3.0.1", DependencyGroup::Default).unwrap();
        package_json
            .add_dependency("fast-decode-uri-component", "1.0.1", DependencyGroup::Dev)
            .unwrap();

        package_json.save().unwrap();

        let package_manager =
            PackageManager::new(&package_json_path, Npmrc::current().leak()).unwrap();
        let args = InstallCommandArgs { prod: false, dev: false, no_optional: false };
        package_manager.install(&args).await.unwrap();

        // Make sure the package is installed
        assert!(dir.path().join("node_modules/is-odd").is_symlink());
        assert!(dir.path().join("node_modules/.pacquet/is-odd@3.0.1").exists());
        // Make sure it installs direct dependencies
        assert!(!dir.path().join("node_modules/is-number").exists());
        assert!(dir.path().join("node_modules/.pacquet/is-number@6.0.0").exists());
        // Make sure we install dev-dependencies as well
        assert!(dir.path().join("node_modules/fast-decode-uri-component").is_symlink());
        assert!(dir.path().join("node_modules/.pacquet/fast-decode-uri-component@1.0.1").is_dir());

        insta::assert_debug_snapshot!(get_all_folders(dir.path()));

        env::set_current_dir(&current_directory).unwrap();
    }
}
