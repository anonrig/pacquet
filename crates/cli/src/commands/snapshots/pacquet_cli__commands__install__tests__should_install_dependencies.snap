---
source: crates/cli/src/commands/install.rs
expression: get_all_folders(&dir.path().to_path_buf())
---
[
    "node_modules",
    "node_modules/.pacquet",
    "node_modules/.pacquet/fast-decode-uri-component@1.0.1",
    "node_modules/.pacquet/fast-decode-uri-component@1.0.1/node_modules",
    "node_modules/.pacquet/fast-decode-uri-component@1.0.1/node_modules/fast-decode-uri-component",
    "node_modules/.pacquet/is-even@1.0.0",
    "node_modules/.pacquet/is-even@1.0.0/node_modules",
    "node_modules/.pacquet/is-even@1.0.0/node_modules/is-even",
    "node_modules/.pacquet/is-even@1.0.0/node_modules/is-odd",
    "node_modules/.pacquet/is-number@6.0.0",
    "node_modules/.pacquet/is-number@6.0.0/node_modules",
    "node_modules/.pacquet/is-number@6.0.0/node_modules/is-number",
    "node_modules/.pacquet/is-odd@0.1.2",
    "node_modules/.pacquet/is-odd@0.1.2/node_modules",
    "node_modules/.pacquet/is-odd@0.1.2/node_modules/is-odd",
    "node_modules/.pacquet/is-odd@3.0.1",
    "node_modules/.pacquet/is-odd@3.0.1/node_modules",
    "node_modules/.pacquet/is-odd@3.0.1/node_modules/is-number",
    "node_modules/.pacquet/is-odd@3.0.1/node_modules/is-odd",
    "node_modules/fast-decode-uri-component",
    "node_modules/is-even",
    "node_modules/is-odd",
]
